File: src/TmrOverlay.App/History/SessionHistoryQueryService.cs
using System.Text.Json;
using TmrOverlay.Core.History;

namespace TmrOverlay.App.History;

internal sealed class SessionHistoryQueryService
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    private readonly SessionHistoryOptions _options;

    public SessionHistoryQueryService(SessionHistoryOptions options)
    {
        _options = options;
    }

    public SessionHistoryLookupResult Lookup(HistoricalComboIdentity combo)
    {
        if (!_options.Enabled)
        {
            return SessionHistoryLookupResult.Empty(combo);
        }

        var userAggregate = ReadAggregate(_options.ResolvedUserHistoryRoot, combo);
        var baselineAggregate = _options.UseBaselineHistory
            ? ReadAggregate(_options.ResolvedBaselineHistoryRoot, combo)
            : null;
        return new SessionHistoryLookupResult(combo, userAggregate, baselineAggregate);
    }

    public CarRadarCalibrationLookupResult LookupCarRadarCalibration(HistoricalComboIdentity combo)
    {
        if (!_options.Enabled)
        {
            return CarRadarCalibrationLookupResult.Empty(combo);
        }

        var userAggregate = ReadCarRadarCalibrationAggregate(_options.ResolvedUserHistoryRoot, combo.CarKey);
        var baselineAggregate = _options.UseBaselineHistory
            ? ReadCarRadarCalibrationAggregate(_options.ResolvedBaselineHistoryRoot, combo.CarKey)
            : null;
        return new CarRadarCalibrationLookupResult(combo, userAggregate, baselineAggregate);
    }

    private static HistoricalSessionAggregate? ReadAggregate(string root, HistoricalComboIdentity combo)
    {
        var path = Path.Combine(
            root,
            "cars",
            combo.CarKey,
            "tracks",
            combo.TrackKey,
            "sessions",
            combo.SessionKey,
            "aggregate.json");

        if (!File.Exists(path))
        {
            return null;
        }

        try
        {
            using var stream = File.OpenRead(path);
            var aggregate = JsonSerializer.Deserialize<HistoricalSessionAggregate>(stream, JsonOptions);
            return aggregate?.AggregateVersion == HistoricalDataVersions.AggregateVersion
                ? aggregate
                : null;
        }
        catch
        {
            return null;
        }
    }

    private static HistoricalCarRadarCalibrationAggregate? ReadCarRadarCalibrationAggregate(
        string root,
        string carKey)
    {
        var path = Path.Combine(
            root,
            "cars",
            carKey,
            "radar-calibration.json");

        if (!File.Exists(path))
        {
            return null;
        }

        try
        {
            using var stream = File.OpenRead(path);
            var aggregate = JsonSerializer.Deserialize<HistoricalCarRadarCalibrationAggregate>(stream, JsonOptions);
            return aggregate?.AggregateVersion == HistoricalDataVersions.CarRadarCalibrationAggregateVersion
                ? aggregate
                : null;
        }
        catch
        {
            return null;
        }
    }
}

File: local-mac/TmrOverlayMac/Sources/TmrOverlayMac/History/SessionHistoryQueryService.swift
import Foundation

final class SessionHistoryQueryService {
    private let userHistoryRoot: URL
    private let baselineHistoryRoot: URL?
    private let useBaselineHistory: Bool

    init(
        userHistoryRoot: URL,
        baselineHistoryRoot: URL? = nil,
        useBaselineHistory: Bool = false
    ) {
        self.userHistoryRoot = userHistoryRoot
        self.baselineHistoryRoot = baselineHistoryRoot
        self.useBaselineHistory = useBaselineHistory
    }

    func lookup(_ combo: HistoricalComboIdentity) -> SessionHistoryLookupResult {
        let userAggregate = readAggregate(root: userHistoryRoot, combo: combo)
        let baselineAggregate = useBaselineHistory
            ? baselineHistoryRoot.flatMap { readAggregate(root: $0, combo: combo) }
            : nil

        return SessionHistoryLookupResult(
            combo: combo,
            userAggregate: userAggregate,
            baselineAggregate: baselineAggregate
        )
    }

    private func readAggregate(root: URL, combo: HistoricalComboIdentity) -> HistoricalSessionAggregate? {
        let url = root
            .appendingPathComponent("cars", isDirectory: true)
            .appendingPathComponent(combo.carKey, isDirectory: true)
            .appendingPathComponent("tracks", isDirectory: true)
            .appendingPathComponent(combo.trackKey, isDirectory: true)
            .appendingPathComponent("sessions", isDirectory: true)
            .appendingPathComponent(combo.sessionKey, isDirectory: true)
            .appendingPathComponent("aggregate.json")

        guard let data = try? Data(contentsOf: url) else {
            return nil
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try? decoder.decode(HistoricalSessionAggregate.self, from: data)
    }
}

struct SessionHistoryLookupResult {
    let combo: HistoricalComboIdentity
    let userAggregate: HistoricalSessionAggregate?
    let baselineAggregate: HistoricalSessionAggregate?

    var preferredAggregate: HistoricalSessionAggregate? {
        userAggregate ?? baselineAggregate
    }

    var preferredAggregateSource: String? {
        if userAggregate != nil {
            return "user"
        }

        if baselineAggregate != nil {
            return "baseline"
        }

        return nil
    }

    var hasAnyData: Bool {
        preferredAggregate != nil
    }

    static func empty(_ combo: HistoricalComboIdentity) -> SessionHistoryLookupResult {
        SessionHistoryLookupResult(combo: combo, userAggregate: nil, baselineAggregate: nil)
    }
}
